lib/hunter/udp_moose.rb

```diff
@@ -0,0 +1,583 @@
+# frozen_string_literal: true
+
+require 'socket'
+require 'json'
+
+class UDPMoose
+  attr_reader :buffers, :requests
+
+  def initialize(port = 0, ingress = false)
+    @udp_socket = UDPSocket.new
+    @udp_socket.setsockopt(Socket::SOL_SOCKET, Socket::SO_BROADCAST, true)
+    @udp_socket.bind('0.0.0.0', port)
+
+    @ingress = ingress
+
+    @requests = {}
+    @buffers = {}
+
+    @receive_lock = UDPMooseLock.new
+
+    @receives = []
+
+    listener
+    timer
+  end
+
+  #
+  # client methods
+  #
+  #
+  #
+
+  # user send request
+  def send(body, server_ip_range, server_port, max_connection = 1, max_timeout = 10)
+    # build the request
+    request_id = ((Time.now.to_f * 1000).to_i.to_s + body).hash.to_s
+    request = UDPMooseRequest.new(request_id, body, server_ip_range, server_port, max_connection, max_timeout)
+
+    @requests[request_id] = request
+
+    Thread.new do
+      # enqueue the initial handshake request connection
+      request.enqueue_by_connection_key(request_id)
+
+      # loop until request end
+      while request.requesting?
+
+        connection = request.dequeue
+        @udp_socket.send(connection.payload, 0, connection.server_ip, request.server_port)
+
+        request.lock_request
+        request.wait_request if request.idle? && request.requesting?
+        request.unlock_request
+
+      end
+    end
+
+    # return the request id for user access in the future
+    request_id
+  end
+
+  def get_responses(request_id)
+    request = @requests[request_id]
+    request.lock_response
+    request.wait_response if request.requesting?
+    request.unlock_response
+
+    responses = [].tap do |arr|
+      request.connections.each do |key, connection|
+        arr << connection.server_ip if connection.closed? && key != request_id
+      end
+    end
+
+    yield responses if block_given?
+  end
+
+  #
+  # server methods
+  #
+  #
+  #
+
+  def receive
+    loop do
+      lock_receive
+      wait_receive if @receives.empty?
+      unlock_receive
+
+      if block_given?
+        buffer = @receives.shift
+        yield buffer.client_ip, buffer.buffer
+      end
+    end
+  end
+
+  private
+
+  # create a thread for listening to the income packets
+  def listener
+    Thread.new do
+      loop do
+        msg, addr = @udp_socket.recvfrom(64_000) # thread blocked until new packet recieved
+        hash_msg = JSON.parse(msg)
+
+        # handshake request received
+        if @ingress && hash_msg['type'] == 'stu'
+
+          unless @buffers[hash_msg['id']]
+            @buffers[hash_msg['id']] =
+              UDPMooseBuffer.new(hash_msg['total'], addr[2], addr[1])
+          end
+          jack(hash_msg['id']) unless @buffers[hash_msg['id']]&.packet&.> 0
```

Comment:
The income packet might be either a packet from client or a packet from server. The `listener` method will parse both of them. 'stu' and 'jack' are handshake packets, the former is sent by the client and the latter is sent by the server back to the client to let the client know that the connection is established. After that, the client will start to send 'request' packets which includes the fragments of real payload, and the server will response with the 'response' packets unless the 'request' packet the server received is the last packet. When the server received the last packet, it will send an 'ivan' packet to the client to start the handwave process and the client will finally send a 'bye' packet. For the refactoring you asked for, could you please give a more detailed requirement? Or do you need some protocol description comments in that file?

Reply:
I think just renaming the packet types to something that better describes the purpose of the packet is what I'm looking for. When I see the code describing a packet with type `jack`, the word `jack` doesn't do a lot to tell me what the packet is actually for. If I see it describing a packet with type `client_handshake`, I have a much better idea of its responsibility and what the following code (i.e. `UDPMoose#jack`) will be doing.